### ISSUE TITLE
Add support for Contour Shuttle

### ISSUE THREAD

**Issue by peternewman** [bitfocus/companion#2430]
### Is this a feature relevant to companion itself, and not a module?

- [X] I believe this to be a feature for companion, not a module

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the feature

Add support for https://contour-design.co.uk/collections/multimedia-controller/products/multimedia-controller-pro and similar
https://contour-design.co.uk/collections/multimedia-controller/products/multimedia-controller-xpress

I assume you'd accept a PR, looking at xkeys it doesn't look too hard... (famous last words).

### Usecases

Let people with Shuttles use Companion.

**Comment by Julusian** [comment]
Yeah, we will happily accept a PR for other devices as long as they can translate across in a not too confusing manner.

Make sure to base your work off the beta branch, and yes use any of the other types as a reference. Be aware the xkeys implementation is a tiny bit special in that it does some meddling to be able to have a single surface span multiple pages.

Is there an existing npm library for these that you intend to use? The HID protocol looks simple enough https://github.com/legacymachine/node-red-contrib-shuttlexpress/blob/master/shuttlexpress/shuttlexpress.js

I am a little concerned on how the shuttle wheel should be mapped. The jog portion is 'simple' and can map just like the rotary encoders of a streamdeck+ or loupedeck, but the only other shuttle is the xkeys which does it as a variable and has very little support for being used anywhere. Perhaps doing that would be good enough for now, as it could be handled pretty easily with some triggers.

**Comment by peternewman** [comment]
> Be aware the xkeys implementation is a tiny bit special in that it does some meddling to be able to have a single surface span multiple pages.

Ah I hadn't realised, hopefully that's fairly easy to spot.
 
> Is there an existing npm library for these that you intend to use? The HID protocol looks simple enough https://github.com/legacymachine/node-red-contrib-shuttlexpress/blob/master/shuttlexpress/shuttlexpress.js

I'd not really looked at all yet, I'd just got as far as the fact that evtest ( https://gitlab.freedesktop.org/libevdev/evtest ) would quite happily show me all the data, so I assume it shouldn't be too hard.

In fact I've just found this (although it doesn't seem to be on npm yet...)
https://github.com/hopejr/ShuttleControlUSB

> I am a little concerned on how the shuttle wheel should be mapped. The jog portion is 'simple' and can map just like the rotary encoders of a streamdeck+ or loupedeck, but the only other shuttle is the xkeys which does it as a variable and has very little support for being used anywhere. Perhaps doing that would be good enough for now, as it could be handled pretty easily with some triggers.

Ah, I didn't realise that was as unique and kind of assumed the Streamdeck device had both too. Presumably matching the xkeys makes sense for now, then either someone can improve general handling of them or transition both or whatever is most appropriate?

**Comment by Julusian** [comment]
> In fact I've just found this (although it doesn't seem to be on npm yet...)
> [hopejr/ShuttleControlUSB](https://github.com/hopejr/ShuttleControlUSB)

Having a very quick skim read of that library, I don't think it will work for us.  

1) It is using a deprecated npm library (usb-detection) that is fiddly to get working in our builds and is prone to crashing, especially on linux. (from experience, we used to use it in companion-satellite)
2) The connection logic is written around assuming only a single shuttle will be attached to a system at a time. https://github.com/hopejr/ShuttleControlUSB/blob/ce3dfd4e2d9378bfefd80be207cd6ca491dffa84/lib/Shuttle.js#L23-L35 In practise may not be a problem, but at some point someone will want to use multiple. 
  And it will be fighting the way we manage usb devices. It could work, but it will be different to how we are doing every other usb device type.

But it can definitely be a reference, or could be forked to remove both of these issues.


> Ah, I didn't realise that was as unique and kind of assumed the Streamdeck device had both too. Presumably matching the xkeys makes sense for now, then either someone can improve general handling of them or transition both or whatever is most appropriate?

Yeah I think for the shuttle portion do it the same as the xkeys. For the jog portion, it would be good to have it the same as the streamdeck/loupedeck.

**Comment by hopejr** [comment]
I'm the author of ShuttleControlUSB. The usb-detection module has only recently be deprecated, and I wasn't aware. Good to know now. I'm happy to make the changes you're needing if that will help, including adding support for multiple shuttles.

**Comment by Julusian** [comment]
@hopejr even better! thanks for the help :)

> The usb-detection module has only recently be deprecated, and I wasn't aware.

Yeah, thats fair. It feels like ages ago to me, as I was working towards that happening for over a year.
 It should be pretty quick and easy to migrate across to `node-usb` https://github.com/node-usb/node-usb#migrating-from-node-usb-detection

>  I'm happy to make the changes you're needing if that will help, including adding support for multiple shuttles

I think that the ideal thing would be to change the class to something like this:
```
class Shuttle extends EventEmitter {
  constructor (watchUsb) {
    super()

    this._watchUsb = watchUsb

	...
  }

  start () {
	if (!this._watchUsb) return

   ...
  }

  stop () {
	if (!this._watchUsb) return

   ...
  }

_connect () {
  // your existing code to find the hid device

  this.openPath(...)
}


openPath(path) {
  // most of your logic from _connect

}
}
```
That should give us what we need, a way to disable the monitoring and to instead instruct it to open for a specific path, and will be backwards compatible.  
Figuring out the vid and pid for the path supplied isnt the simplest, and does require doing a `HID.devices()` call and looking through the results for a match. I am working upstream on getting this improved.

An unrelated thought, do these devices have serial numbers? If they do then the `HID.devices()` call will report them. The lack of unique id is a bit of a pain for xkeys, but streamdecks and loupedecks have them which makes supporting multiple on a single machine so much predictable

**Comment by hopejr** [comment]
@Julusian I've just updated it to use usb instead of usb-detection. It was a very simple change to make.

Unfortunately, the devices don't have serial numbers. I have an idea for multiple devices to work regardless of this though. At the moment I only have one device with me, but I can grab one from work tomorrow to test more than one. It would work with the current events and wouldn't need to use paths in the way you suggested.

**Comment by hopejr** [comment]
@Julusian I've just committed a multiple device version of ShuttleControlUSB.

**Comment by Julusian** [comment]
@hopejr I feel like that might fight how we are managing the other usb devices, but it might be fine.  
One bit that does concern me is that we have a setting to toggle watching for usb devices, which will change everything except for these devices. But they are probably going to be uncommon enough that its probably not going to be an issue in practise.
And we are currently running each usb device in its own forked process, for reasons I am not sure of, but havent gotten around to changing. I suspect because the current versions of node-hid are only able to do 3 devices reliably at once. That separation will probably add some complications here, to make sure that system supports multiple per forked process. Or perhaps simpler is to bypass that wrapping for this, as that limit is unlikely to be an issue

@peternewman hmm yeah there isnt really an up to date guide there.
The short version is:
1) `sudo apt install libusb-1.0-0-dev libudev-dev` to install the usb system dependencies
1) Clone the repository and checkout the `beta` branch. 
1) Install nodejs 18. I recommend using a version manager to do this such as https://github.com/Schniz/fnm or https://github.com/nvm-sh/nvm. You can install one of them, then in the git clone run `fnm use` or `nvm use` and it should install the correct version
2) You may need to install `yarn`, which might be possible with `corepack enable` or `sudo npm install -g yarn`
3) Run `yarn update` to prepare various bits of the application (install dependencies, build webui)
4) Run it with `yarn dev`
5) (optional) if you want to develop the webui too, in another shell run `yarn dev-webui`

**Comment by peternewman** [comment]
I've currently got buttons working with the device. There's an offset of one, but I've not really tried to track it down yet.

> @hopejr I feel like that might fight how we are managing the other usb devices, but it might be fine.

The openPath thing that @Julusian mentioned would certainly be useful. The device gets detected fine by the common HID code that already exists, and triggers creating it, but I then can't get the model name info, as it's not exposed from your library and I can't open or query a particular path. I'm currently hackily getting the first device from getDeviceList(). Also because the path isn't returned in that info, I can't match it there either.

> @peternewman hmm yeah there isnt really an up to date guide there. The short version is:

That seemed to generally work thanks @Julusian ! Although it felt like it was doing some production builds at times which I didn't need, but maybe there's no way to skip them.

**Comment by hopejr** [comment]
> Exposing some of the static VID/PID/Name info as constants/an array would also be nice if possible @hopejr .

I've just added a new function to retrieve the raw HID device object. Just need to supply the id as a parameter.

**Comment by peternewman** [comment]
See https://github.com/bitfocus/companion/pull/2436 reviews welcome!

@hopejr I was hoping for access by path.

Having the VID and PID as constants would be nice too.

**Comment by hopejr** [comment]
@peternewman could you do a PR on shuttle-control-usb with what you're needing?

**Comment by peternewman** [comment]
> @peternewman could you do a PR on shuttle-control-usb with what you're needing?

https://github.com/hopejr/ShuttleControlUSB/pull/5 covers part of it.

I've opened https://github.com/hopejr/ShuttleControlUSB/issues/6 to try and better explain and track the other request as I'm not really sure of the right syntax to use/preferred way to do it.

**Comment by hopejr** [comment]
@peternewman and I have made changes to ShuttleControlUSB that should help with this issue. It's been released as version 1.3.0.
